File: astroberrymanager/almanac.py
#!/usr/bin/env python3
# coding=utf-8

"""
Copyright(c) 2026 Radek Kaczorek  <rkaczorek AT gmail DOT com>

This library is part of Astroberry OS and Astroberry Manager
https://github.com/astroberry-official/astroberry-os
https://github.com/astroberry-official/astroberry-manager

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License version 3 as published by the Free Software Foundation.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public License
along with this library; see the file COPYING.LIB.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA.
"""

import datetime, ephem, numpy

def getAlmanac(socketio, gpstime, latitude, longitude, elevation):
    t = datetime.datetime.strptime(str(gpstime), '%Y-%m-%dT%H:%M:%S.%f%z')
    t = t.replace(tzinfo=datetime.timezone.utc) #Convert it to an aware datetime object in UTC time.

    lat = "%s" % latitude
    lon = "%s" % longitude
    alt = "%.2f" % elevation

    # init observer
    home = ephem.Observer()

    # set geo position
    home.lat = lat
    home.lon = lon
    home.elevation = float(alt)
    home.date = t

    # get polaris data
    polaris_data = getPolarisData(home)

    # emit celestial data
    socketio.emit('almanac', {
    'latitude': "%.2f" % numpy.degrees(home.lat),
    'longitude': "%.2f" % numpy.degrees(home.lon),
    'elevation': "%.2f" % home.elevation,
    'polaris_hour_angle': "%.2f" % polaris_data[0],
    'polaris_next_transit': "%s" % polaris_data[1],
    'moon_phase': "%s" % getMoonPhase(home),
    'moon_light': "%.1f" % ephem.Moon(home).phase,
    'moon_rise': "%s" % getBodyPositions(home,ephem.Moon(home))[0],
    'moon_transit': "%s" % getBodyPositions(home,ephem.Moon(home))[1],
    'moon_set': "%s" % getBodyPositions(home,ephem.Moon(home))[2],
    'moon_az': "%.2f°" % numpy.degrees(ephem.Moon(home).az),
    'moon_alt': "%.2f°" % numpy.degrees(ephem.Moon(home).alt),
    'moon_ra': "%.2f" % numpy.degrees(ephem.Moon(home).ra),
    'moon_dec': "%.2f" % numpy.degrees(ephem.Moon(home).dec),
    'moon_new': "%s" % ephem.localtime(ephem.next_new_moon(t)).strftime("%Y-%m-%d %H:%M:%S"),
    'moon_full': "%s" % ephem.localtime(ephem.next_full_moon(t)).strftime("%Y-%m-%d %H:%M:%S"),
    'sun_at_start': getSunTwilights(home)[2][0],
    'sun_ct_start': getSunTwilights(home)[0][0],
    'sun_rise': "%s" % getBodyPositions(home,ephem.Sun(home))[0],
    'sun_transit': "%s" % getBodyPositions(home,ephem.Sun(home))[1],
    'sun_set': "%s" % getBodyPositions(home,ephem.Sun(home))[2],
    'sun_ct_end': getSunTwilights(home)[0][1],
    'sun_at_end': getSunTwilights(home)[2][1],
    'sun_az': "%.2f°" % numpy.degrees(ephem.Sun(home).az),
    'sun_alt': "%.2f°" % numpy.degrees(ephem.Sun(home).alt),
    'sun_ra': "%.2f" % numpy.degrees(ephem.Sun(home).ra),
    'sun_dec': "%.2f" % numpy.degrees(ephem.Sun(home).dec),
    'sun_equinox': "%s" % ephem.localtime(ephem.next_equinox(t)).strftime("%Y-%m-%d %H:%M:%S"),
    'sun_solstice': "%s" % ephem.localtime(ephem.next_solstice(t)).strftime("%Y-%m-%d %H:%M:%S"),
    'mercury_rise': "%s" % getBodyPositions(home,ephem.Mercury(home))[0],
    'mercury_transit': "%s" % getBodyPositions(home,ephem.Mercury(home))[1],
    'mercury_set': "%s" % getBodyPositions(home,ephem.Mercury(home))[2],
    'mercury_az': "%.2f°" % numpy.degrees(ephem.Mercury(home).az),
    'mercury_alt': "%.2f°" % numpy.degrees(ephem.Mercury(home).alt),
    'venus_rise': "%s" % getBodyPositions(home,ephem.Venus(home))[0],
    'venus_transit': "%s" % getBodyPositions(home,ephem.Venus(home))[1],
    'venus_set': "%s" % getBodyPositions(home,ephem.Venus(home))[2],
    'venus_az': "%.2f°" % numpy.degrees(ephem.Venus(home).az),
    'venus_alt': "%.2f°" % numpy.degrees(ephem.Venus(home).alt),
    'mars_rise': "%s" % getBodyPositions(home,ephem.Mars(home))[0],
    'mars_transit': "%s" % getBodyPositions(home,ephem.Mars(home))[1],
    'mars_set': "%s" % getBodyPositions(home,ephem.Mars(home))[2],
    'mars_az': "%.2f°" % numpy.degrees(ephem.Mars(home).az),
    'mars_alt': "%.2f°" % numpy.degrees(ephem.Mars(home).alt),
    'jupiter_rise': "%s" % getBodyPositions(home,ephem.Jupiter(home))[0],
    'jupiter_transit': "%s" % getBodyPositions(home,ephem.Jupiter(home))[1],
    'jupiter_set': "%s" % getBodyPositions(home,ephem.Jupiter(home))[2],
    'jupiter_az': "%.2f°" % numpy.degrees(ephem.Jupiter(home).az),
    'jupiter_alt': "%.2f°" % numpy.degrees(ephem.Jupiter(home).alt),
    'saturn_rise': "%s" % getBodyPositions(home,ephem.Saturn(home))[0],
    'saturn_transit': "%s" % getBodyPositions(home,ephem.Saturn(home))[1],
    'saturn_set': "%s" % getBodyPositions(home,ephem.Saturn(home))[2],
    'saturn_az': "%.2f°" % numpy.degrees(ephem.Saturn(home).az),
    'saturn_alt': "%.2f°" % numpy.degrees(ephem.Saturn(home).alt),
    'uranus_rise': "%s" % getBodyPositions(home,ephem.Uranus(home))[0],
    'uranus_transit': "%s" % getBodyPositions(home,ephem.Uranus(home))[1],
    'uranus_set': "%s" % getBodyPositions(home,ephem.Uranus(home))[2],
    'uranus_az': "%.2f°" % numpy.degrees(ephem.Uranus(home).az),
    'uranus_alt': "%.2f°" % numpy.degrees(ephem.Uranus(home).alt),
    'neptune_rise': "%s" % getBodyPositions(home,ephem.Neptune(home))[0],
    'neptune_transit': "%s" % getBodyPositions(home,ephem.Neptune(home))[1],
    'neptune_set': "%s" % getBodyPositions(home,ephem.Neptune(home))[2],
    'neptune_az': "%.2f°" % numpy.degrees(ephem.Neptune(home).az),
    'neptune_alt': "%.2f°" % numpy.degrees(ephem.Neptune(home).alt)
    })

def getMoonPhase(observer):
    target_date_utc = observer.date
    target_date_local = ephem.localtime( target_date_utc ).date()
    next_full = ephem.localtime( ephem.next_full_moon(target_date_utc) ).date()
    next_new = ephem.localtime( ephem.next_new_moon(target_date_utc) ).date()
    next_last_quarter = ephem.localtime( ephem.next_last_quarter_moon(target_date_utc) ).date()
    next_first_quarter = ephem.localtime( ephem.next_first_quarter_moon(target_date_utc) ).date()
    previous_full = ephem.localtime( ephem.previous_full_moon(target_date_utc) ).date()
    previous_new = ephem.localtime( ephem.previous_new_moon(target_date_utc) ).date()
    previous_last_quarter = ephem.localtime( ephem.previous_last_quarter_moon(target_date_utc) ).date()
    previous_first_quarter = ephem.localtime( ephem.previous_first_quarter_moon(target_date_utc) ).date()

    if target_date_local in (next_full, previous_full):
        return 'Full'
    elif target_date_local in (next_new, previous_new):
        return 'New'
    elif target_date_local in (next_first_quarter, previous_first_quarter):
        return 'First Quarter'
    elif target_date_local in (next_last_quarter, previous_last_quarter):
        return 'Last Quarter'
    elif previous_new < next_first_quarter < next_full < next_last_quarter < next_new:
        return 'Waxing Crescent'
    elif previous_first_quarter < next_full < next_last_quarter < next_new < next_first_quarter:
        return 'Waxing Gibbous'
    elif previous_full < next_last_quarter < next_new < next_first_quarter < next_full:
        return 'Waning Gibbous'
    elif previous_last_quarter < next_new < next_first_quarter < next_full < next_last_quarter:
        return 'Waning Crescent'

def getBodyPositions(observer, body):
    positions = []

    # test for always below horizon or always above horizon
    try:
        if ephem.localtime(observer.previous_rising(body)).date() == ephem.localtime(observer.date).date() and observer.previous_rising(body) < observer.previous_transit(body) < observer.previous_setting(body) < observer.date:
            positions.append(observer.previous_rising(body))
            positions.append(observer.previous_transit(body))
            positions.append(observer.previous_setting(body))
        elif ephem.localtime(observer.previous_rising(body)).date() == ephem.localtime(observer.date).date() and observer.previous_rising(body) < observer.previous_transit(body) < observer.date < observer.next_setting(body):
            positions.append(observer.previous_rising(body))
            positions.append(observer.previous_transit(body))
            positions.append(observer.next_setting(body))
        elif ephem.localtime(observer.previous_rising(body)).date() == ephem.localtime(observer.date).date() and observer.previous_rising(body) < observer.date < observer.next_transit(body) < observer.next_setting(body):
            positions.append(observer.previous_rising(body))
            positions.append(observer.next_transit(body))
            positions.append(observer.next_setting(body))
        elif ephem.localtime(observer.previous_rising(body)).date() == ephem.localtime(observer.date).date() and  observer.date < observer.next_rising(body) < observer.next_transit(body) < observer.next_setting(body):
            positions.append(observer.next_rising(body))
            positions.append(observer.next_transit(body))
            positions.append(observer.next_setting(body))
        else:
            positions.append(observer.next_rising(body))
            positions.append(observer.next_transit(body))
            positions.append(observer.next_setting(body))
    except (ephem.NeverUpError, ephem.AlwaysUpError):
        try:
            if ephem.localtime(observer.previous_transit(body)).date() == ephem.localtime(observer.date).date() and observer.previous_transit(body) < observer.date:
                positions.append('-')
                positions.append(observer.previous_transit(body))
                positions.append('-')
            elif ephem.localtime(observer.previous_transit(body)).date() == ephem.localtime(observer.date).date() and observer.next_transit(body) > observer.date:
                positions.append('-')
                positions.append(observer.next_transit(body))
                positions.append('-')
            else:
                positions.append('-')
                positions.append('-')
                positions.append('-')
        except (ephem.NeverUpError, ephem.AlwaysUpError):
            positions.append('-')
            positions.append('-')
            positions.append('-')

    if positions[0] != '-':
        positions[0] = ephem.localtime( positions[0] ).strftime("%H:%M:%S")
    if positions[1] != '-':
        positions[1] = ephem.localtime( positions[1] ).strftime("%H:%M:%S")
    if positions[2] != '-':
        positions[2] = ephem.localtime( positions[2] ).strftime("%H:%M:%S")

    return positions

def getSunTwilights(observer):
    results = []

    """
    An observer at the North Pole would see the Sun circle the sky at 23.5° above the horizon all day.
    An observer at 90° – 23.5° = 66.5° would see the Sun spend the whole day on the horizon, making a circle along its circumference.
    An observer would have to be at 90° – 23.5° – 18° = 48.5° latitude or even further south in order for the Sun to dip low enough for them to observe the level of darkness defined as astronomical twilight. 

    civil twilight = -6
    nautical twilight = -12
    astronomical twilight = -18

    getSunTwilights(home)[0][0]   -   civil twilight end
    getSunTwilights(home)[0][1]   -   civil twilight start

    getSunTwilights(home)[1][0]   -   nautical twilight end
    getSunTwilights(home)[1][1]   -   nautical twilight start

    getSunTwilights(home)[2][0]   -   astronomical twilight end
    getSunTwilights(home)[2][1]   -   astronomical twilight start
    """

    # remember entry observer horizon
    observer_horizon = observer.horizon

    # twilights, their horizons and whether to use the centre of the Sun or not
    twilights = [('-6', True), ('-12', True), ('-18', True)]

    for twi in twilights:
        observer.horizon = twi[0]
        try:
            rising_setting = getBodyPositions(observer,ephem.Sun(observer))
            results.append((rising_setting[0], rising_setting[2]))
        except ephem.AlwaysUpError:
            results.append(('n/a', 'n/a'))

    # reset observer horizon to entry
    observer.horizon = observer_horizon

    return results

def getPolarisData(observer):
    polaris_data = []

    """
    lst = 100.46 + 0.985647 * d + lon + 15 * ut [based on http://www.stargazing.net/kepler/altaz.html]
    d - the days from J2000 (1200 hrs UT on Jan 1st 2000 AD), including the fraction of a day
    lon - your longitude in decimal degrees, East positive
    ut - the universal time in decimal hours
    """

    j2000 = ephem.Date('2000/01/01 12:00:00')
    d = observer.date - j2000

    lon = numpy.rad2deg(float(observer.lon))

    utstr = observer.date.datetime().strftime("%H:%M:%S")
    ut = float(utstr.split(":")[0]) + float(utstr.split(":")[1])/60 + float(utstr.split(":")[2])/3600

    lst = 100.46 + 0.985647 * d + lon + 15 * ut
    lst = lst - int(lst / 360) * 360

    polaris = ephem.readdb("Polaris,f|M|F7,2:31:49.095,89:15:50.79,2.02,2000")
    polaris.compute()

    polaris_ra_deg = numpy.rad2deg(float(polaris._ra))

    # polaris Hour Angle = LST - RA Polaris [expressed in degrees or 15*(h+m/60+s/3600)]
    pha = lst - polaris_ra_deg

    # normalize
    if pha < 0:
        pha += 360
    elif pha > 360:
        pha -= 360

    # append polaris hour angle
    polaris_data.append(pha)

    # append polaris next transit
    try:
        pnt = ephem.localtime(observer.next_transit(polaris)).strftime("%H:%M:%S")
        polaris_data.append(pnt)
    except (ephem.NeverUpError, ephem.AlwaysUpError):
        polaris_data.append('-')

    return polaris_data

File: astroberrymanager/assets/js/app/almanac.js
/*
 Copyright(c) 2026 Radek Kaczorek  <rkaczorek AT gmail DOT com>

 This library is part of Astroberry OS and Astroberry Manager
 https://github.com/astroberry-official/astroberry-os
 https://github.com/astroberry-official/astroberry-manager

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Library General Public
 License version 3 as published by the Free Software Foundation.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Library General Public License for more details.

 You should have received a copy of the GNU Library General Public License
 along with this library; see the file COPYING.LIB.  If not, write to
 the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 Boston, MA 02110-1301, USA.
*/

import { geoLocation } from './location.js';
import { centerOnSolarObject } from './celestial.js';
import { updateMoon } from './moon.js';
import { syslogPrint } from './helpers.js';
import { deg2dms, deg2hms } from './functions.js';
import { socket } from './sockets.js';

function updateAlmanac(data) {
    if (!data) return;

    // Moon
    $("#moon_phase_name").html(data.moon_phase);
    $("#moon_light").html(data.moon_light + "%");
    $("#moon_rise").html(data.moon_rise);
    $("#moon_transit").html(data.moon_transit);
    $("#moon_set").html(data.moon_set);
    $("#moon_az").html(deg2dms(parseFloat(data.moon_az)));
    $("#moon_alt").html(deg2dms(parseFloat(data.moon_alt)));
    $("#moon_ra").html(deg2hms(data.moon_ra));
    $("#moon_dec").html(deg2dms(data.moon_dec));
    $(".moon_new").html(data.moon_new);
    $("#moon_full").html(data.moon_full);

    // Sun
    $("#sun_at_start").html(data.sun_at_start);
    $("#sun_ct_start").html(data.sun_ct_start);
    $("#sun_rise").html(data.sun_rise);
    $("#sun_transit").html(data.sun_transit);
    $("#sun_set").html(data.sun_set);
    $("#sun_ct_end").html(data.sun_ct_end);
    $("#sun_at_end").html(data.sun_at_end);
    $("#sun_az").html(deg2dms(parseFloat(data.sun_az)));
    $("#sun_alt").html(deg2dms(parseFloat(data.sun_alt)));
    $("#sun_ra").html(deg2hms(data.sun_ra));
    $("#sun_dec").html(deg2dms(data.sun_dec));
    $("#sun_equinox").html(data.sun_equinox);
    $(".sun_solstice").html(data.sun_solstice);

    // Mercury
    $("#mercury_rise").html(data.mercury_rise);
    $("#mercury_transit").html(data.mercury_transit);
    $("#mercury_set").html(data.mercury_set);
    $("#mercury_az").html(data.mercury_az);
    toggleAltTrend("mercury", data.mercury_alt); // set alt trend
    $("#mercury_alt").html(data.mercury_alt);

    // Venus
    $("#venus_rise").html(data.venus_rise);
    $("#venus_transit").html(data.venus_transit);
    $("#venus_set").html(data.venus_set);
    $("#venus_az").html(data.venus_az);
    toggleAltTrend("venus", data.venus_alt); // set alt trend
    $("#venus_alt").html(data.venus_alt);

    // Mars
    $("#mars_rise").html(data.mars_rise);
    $("#mars_transit").html(data.mars_transit);
    $("#mars_set").html(data.mars_set);
    $("#mars_az").html(data.mars_az);
    toggleAltTrend("mars", data.mars_alt); // set alt trend
    $("#mars_alt").html(data.mars_alt);

    // Jupiter
    $("#jupiter_rise").html(data.jupiter_rise);
    $("#jupiter_transit").html(data.jupiter_transit);
    $("#jupiter_set").html(data.jupiter_set);
    $("#jupiter_az").html(data.jupiter_az);
    toggleAltTrend("jupiter", data.jupiter_alt); // set alt trend
    $("#jupiter_alt").html(data.jupiter_alt);

    // Saturn
    $("#saturn_rise").html(data.saturn_rise);
    $("#saturn_transit").html(data.saturn_transit);
    $("#saturn_set").html(data.saturn_set);
    $("#saturn_az").html(data.saturn_az);
    toggleAltTrend("saturn", data.saturn_alt); // set alt trend
    $("#saturn_alt").html(data.saturn_alt);

    // Uranus
    $("#uranus_rise").html(data.uranus_rise);
    $("#uranus_transit").html(data.uranus_transit);
    $("#uranus_set").html(data.uranus_set);
    $("#uranus_az").html(data.uranus_az);
    toggleAltTrend("uranus", data.uranus_alt); // set alt trend
    $("#uranus_alt").html(data.uranus_alt);

    // Neptune
    $("#neptune_rise").html(data.neptune_rise);
    $("#neptune_transit").html(data.neptune_transit);
    $("#neptune_set").html(data.neptune_set);
    $("#neptune_az").html(data.neptune_az);
    toggleAltTrend("neptune", data.neptune_alt); // set alt trend
    $("#neptune_alt").html(data.neptune_alt);

    /* Format polaris hour angle for display */
    var pha = data.polaris_hour_angle;
    $("#pha").html(deg2hms(pha));

    /* Set polaris marker */
    getReticle(pha * Math.PI / 180);

    /* Compute polaris cardinal positions */
    var pnt = data.polaris_next_transit.split(':');
    $("#polaris_next_transit").html(("0" + pnt[0]).substr(-2) + "ʰ " + pnt[1] + "ᵐ " + pnt[2] + "ˢ");

    /* Color-code planets above horizon */
    var planets = ["mercury", "venus", "mars", "jupiter", "saturn", "uranus", "neptune"]
    $.each(planets, function (index, planet) {
        var planet_alt;
        eval('planet_alt = parseFloat(data.' + planet + '_alt' + ');'); // create dynamic variable
        // set color based on altitude
        $("#" + planet).removeClass(); // remove existing color
        if ( planet_alt > 25 ) { // set color based on altitude
            $("#" + planet).addClass("highhorizon");
        } else if ( planet_alt > 0 ) {
            $("#" + planet).addClass("lowhorizon");
        } else {
            $("#" + planet).addClass("belowhorizon");
        }
    });

    /* Set  Equinox and Solstice in order */
    var ss = new Date(data.sun_solstice);
    var se = new Date(data.sun_equinox);
    if (ss < se) { 
        $("#sun_solstice_first").css({display: ""});
        $("#sun_solstice_second").css({display: "none"});
    } else {
        $("#sun_solstice_first").css({display: "none"});
        $("#sun_solstice_second").css({display: ""});
    };

    /* Set  New Moon and Full Moon in order */
    var nm = new Date(data.moon_new);
    var fm = new Date(data.moon_full);
    if (nm < fm) { 
        $("#new_moon_first").css({display: ""});
        $("#new_moon_second").css({display: "none"});
    } else {
        $("#new_moon_first").css({display: "none"});
        $("#new_moon_second").css({display: ""});
    };

    // Update Moon Phase
    var shadow = parseFloat(data.moon_light)/100;
    var phase = data.moon_phase.split(" ")[0];
    if ( phase == "Waning") {
        shadow *= -1
    }
    updateMoon(shadow);

    //console.log("Almanac updated");
}

function getReticle(pha_angle) {
    pha_angle = pha_angle ? pha_angle : 0;
    var width = parseInt($("#polaris-reticle").width());
    var height = parseInt($("#polaris-reticle").height());

    width = 360;
    height = 360;

    if (!width || !height) return;

    var center_x = parseInt(width/2);
    var center_y = parseInt(height/2);
    var radius = parseInt(0.375 * Math.min(width, height));
    var major = parseInt(radius / 7);
    var minor = parseInt(radius / 10);
    var polaris_radius = minor * 0.5;
    var reticle_color = "#40404033";
    var reticle_border_color = "#66666666";
    var grid_color = "#ff3300";
    var text_color = "red";
    var polaris_color = "#ffffcc";
    var text_color_small = "white";
    var text_font = "Roboto Medium";
    var text_font_size = (width > 200) ? parseInt(width/27) : 8;

    const canvas = document.getElementById("polaris-reticle");
    const ctx = canvas.getContext("2d");

    // clear canvas
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    // set lines solid
    ctx.setLineDash([]);
    ctx.lineWidth = 1;

    // Draw reticle
    ctx.beginPath();
    ctx.arc(center_x, center_y, radius, 0, 2 * Math.PI);
    ctx.fillStyle = reticle_border_color;
    ctx.fill();

    ctx.beginPath();
    ctx.arc(center_x, center_y, radius - minor, 0, 2 * Math.PI);
    ctx.fillStyle = reticle_color;
    ctx.fill();

    ctx.beginPath();
    ctx.arc(center_x, center_y, radius - minor/2, 0, 2 * Math.PI);
    ctx.strokeStyle = grid_color;
    ctx.stroke();

    // Draw major ticks
    ctx.lineWidth = 2;
    ctx.beginPath();

    ctx.moveTo(center_x, center_y);
    ctx.lineTo(center_x, center_y - radius - major/2);
    ctx.moveTo(center_x, center_y);
    ctx.lineTo(center_x, center_y + radius + major/2);
    ctx.moveTo(center_x, center_y);
    ctx.lineTo(center_x + radius + major/2, center_y);
    ctx.moveTo(center_x, center_y);
    ctx.lineTo(center_x - radius - major/2, center_y);

    var angle = 360 * Math.PI / 180;
    var step = 45 * Math.PI / 180;

    var x1, y1, x2, y2;

    for (var i = 0; i < angle ; i += step) {
        x1 = center_x + (radius - minor - major/5) * Math.sin(i);
        y1 = center_y + (radius - minor - major/5) * Math.cos(i);
        x2 = center_x + (radius + major/5) * Math.sin(i);
        y2 = center_y + (radius + major/5) * Math.cos(i);

        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
    }

    ctx.strokeStyle = grid_color;
    ctx.stroke();

    // Draw minor ticks
    ctx.lineWidth = 2;
    ctx.beginPath();

    var step = 15 * Math.PI / 180;

    for (i = 0; i < angle ; i += step) {
        x1 = center_x + (radius - minor) * Math.sin(i);
        y1 = center_y + (radius - minor) * Math.cos(i);
        x2 = center_x + (radius) * Math.sin(i);
        y2 = center_y + (radius) * Math.cos(i);

        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
    }

    ctx.strokeStyle = grid_color;
    ctx.stroke();

    ctx.lineWidth = 1;
    ctx.beginPath();

    var step = 5 * Math.PI / 180;

    for (i = 0; i < angle ; i += step) {
        x1 = center_x + (radius - minor + minor/6) * Math.sin(i);
        y1 = center_y + (radius - minor + minor/6) * Math.cos(i);
        x2 = center_x + (radius - minor/6) * Math.sin(i);
        y2 = center_y + (radius - minor/6) * Math.cos(i);

        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
    }

    ctx.strokeStyle = grid_color;
    ctx.stroke();

    // Labels
    ctx.font = text_font_size + "px " + text_font;
    ctx.fillStyle = text_color;
    ctx.fillText("12", center_x - text_font_size/2, center_y - radius - text_font_size);
    ctx.fillText("0", center_x - text_font_size/3, center_y + radius + text_font_size * 1.5);
    ctx.fillText("6", center_x + radius + text_font_size, center_y + text_font_size/4);
    ctx.fillText("18", center_x - radius - text_font_size * 2, center_y + text_font_size/4);

    ctx.font = text_font_size + "px " + text_font;
    ctx.fillStyle = text_color_small;
    ctx.fillText("View as in polar", 5, 15);
    ctx.fillText("finder scope", 5, 15 + text_font_size);


    // draw Polaris legend
    ctx.lineWidth = 2;
    ctx.beginPath();
    ctx.arc(20, height - 50, polaris_radius * 0.8, 0, 2 * Math.PI);
    ctx.strokeStyle = polaris_color;
    ctx.stroke();

    ctx.font = text_font_size + "px " + text_font;
    ctx.fillStyle = grid_color;
    ctx.fillText("Polaris", 20 + text_font_size, height - 45);

    // Draw Polaris
    ctx.beginPath();
    var x = center_x + (radius - minor/2) * Math.sin(pha_angle);
    var y = center_y + (radius - minor/2) * Math.cos(pha_angle);
    ctx.arc(x, y, polaris_radius, 0, 2 * Math.PI);
    ctx.strokeStyle = polaris_color;
    ctx.stroke();

    ctx.lineWidth = 1;

    ctx.beginPath();
    x = center_x + (radius - minor) * Math.sin(pha_angle);
    y = center_y + (radius - minor) * Math.cos(pha_angle);
    ctx.setLineDash([1, 3]); // set lines dashed
    ctx.moveTo(center_x, center_y);
    ctx.lineTo(x, y);
    ctx.strokeStyle = polaris_color;
    ctx.stroke();
}

function renderMoon(phase) { // https://codepen.io/anowodzinski/pen/ZWKXPQ
    phase = 0.5;

    //console.log("Moon phase: " + phase);

    var width = parseInt($("#moon_phase").width());
    var height = parseInt($("#moon_phase").height());

    const canvas = document.getElementById("moon_phase");
    const ctx = canvas.getContext("2d");

    ctx.scale(1, 1);

	const radius = width / 2;

    // clear canvas
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    ctx.fillStyle = '#000000cc';
    //ctx.fillStyle = 'red';

    ctx.beginPath();
    ctx.arc(radius, radius, radius - 5, 0, 2 * Math.PI);
    ctx.closePath();
    ctx.fill();

    ctx.globalCompositeOperation = 'destination-in';

    var c = (5 + phase);

    ctx.beginPath();
    ctx.arc(radius - c * phase * radius * 2, radius, radius * c, 0, 2 * Math.PI);
    ctx.fill();
}

function toggleAltTrend(planet, alt) {
    if (!planet || !alt) return;
    var last_alt = parseFloat($("#" + planet + "_alt").text());
    var new_alt = parseFloat(alt);
    //console.log(new_alt, last_alt);

    // set trend
    if ( new_alt > last_alt) {
        $("#" + planet + " span.planet_trend").removeClass("fa-arrow-down");
        $("#" + planet + " span.planet_trend").addClass("fa-arrow-up");
        $("#" + planet + " span.planet_trend").css("color", "#009933");
    } else if ( new_alt < last_alt) {
        $("#" + planet + " span.planet_trend").removeClass("fa-arrow-up");
        $("#" + planet + " span.planet_trend").addClass("fa-arrow-down");
        $("#" + planet + " span.planet_trend").css("color", "#ff3300");
    }
}

function requestAlmanac() {
    var data = {};
    data['time'] = new Date().toISOString();
    data['latitude'] = parseFloat(geoLocation.latitude);
    data['longitude'] = parseFloat(geoLocation.longitude);
    data['altitude'] = parseFloat(geoLocation.altitude);

    if (!data) return;

    socket.timeout(5000).emit("almanac", data, (err) => {
        if (err) {
            syslogPrint("Almanac data request timed out", "danger");
        } else {
            //syslogPrint("Almanac data requested");
        }
    });
}


/* ================================================================== */
/*                             EVENTS
/* ================================================================== */

function almanacEvents() {
    $("#toggle-moon").on("click", function () {
        toggleMoon();
    });

    $("#toggle-sun").on("click", function () {
        toggleSun();
    });

    $("#toggle-planets").on("click", function () {
        togglePlanets();
    });

    $("#toggle-polaris").on("click", function () {
        togglePolaris();
    });

    $("#moon-actions-center").on("click", function() {
        centerOnSolarObject("lun");
    });

    $("#sun-actions-center").on("click", function() {
        centerOnSolarObject("sol");
    });

    $("#mercury").on("click", function() {
        centerOnSolarObject("mer");
    });

    $("#venus").on("click", function() {
        centerOnSolarObject("ven");
    });

    $("#mars").on("click", function() {
        centerOnSolarObject("mar");
    });

    $("#jupiter").on("click", function() {
        centerOnSolarObject("jup");
    });

    $("#saturn").on("click", function() {
        centerOnSolarObject("sat");
    });

    $("#uranus").on("click", function() {
        centerOnSolarObject("ura");
    });

    $("#neptune").on("click", function() {
        centerOnSolarObject("nep");
    });
}

function toggleMoon() {
    $("#toggle-polaris").removeClass("button-active");
    $("#toggle-moon").addClass("button-active");
    $("#toggle-sun").removeClass("button-active");
    $("#toggle-planets").removeClass("button-active");

    $("#polaris").css({display: "none"});
    $("#moon").css({display: "block"});
    $("#sun").css({display: "none"});
    $("#planets").css({display: "none"});
}

function toggleSun() {
    $("#toggle-polaris").removeClass("button-active");
    $("#toggle-moon").removeClass("button-active");
    $("#toggle-sun").addClass("button-active");
    $("#toggle-planets").removeClass("button-active");

    $("#polaris").css({display: "none"});
    $("#moon").css({display: "none"});
    $("#sun").css({display: "block"});
    $("#planets").css({display: "none"});
}

function togglePlanets() {
    $("#toggle-polaris").removeClass("button-active");
    $("#toggle-moon").removeClass("button-active");
    $("#toggle-sun").removeClass("button-active");
    $("#toggle-planets").addClass("button-active");

    $("#polaris").css({display: "none"});
    $("#moon").css({display: "none"});
    $("#sun").css({display: "none"});
    $("#planets").css({display: "block"});
}

function togglePolaris() {
    $("#toggle-polaris").addClass("button-active");
    $("#toggle-moon").removeClass("button-active");
    $("#toggle-sun").removeClass("button-active");
    $("#toggle-planets").removeClass("button-active");

    $("#polaris").css({display: "block"});
    $("#moon").css({display: "none"});
    $("#sun").css({display: "none"});
    $("#planets").css({display: "none"});
}

export {
    requestAlmanac,
    updateAlmanac,
    getReticle,
    renderMoon,
    toggleAltTrend,
    almanacEvents
};
